File: python/39.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """

        Parameters
        ----------
        candidates : List[int]
            输入的无重复的数组
            
        target : int
            求和的目标值

        Returns
        -------
        List[List[int]]
            结果的组合
        """
        res = []
        
        temp = []
        
        n = len(candidates)
        
        def dfs(_target: int, idx:int):
            if _target < 0:
                return  # 剪枝
            elif _target == 0:
                res.append(temp[:])
                return
        
            for idx in range(idx, n):
                val = candidates[idx]
                if val <= _target:
                    temp.append(val)
                    dfs(_target - val, idx)
                    temp.pop()
                    
        dfs(target, 0)
        return res

File: cpp/39.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> temp;
        dfs(0, target, candidates, temp, res);
        return res;
    }

    void dfs(int idx, int _target, vector<int>& candidates, vector<int>& temp, vector<vector<int>>& res) {

        if (_target == 0) {
            res.emplace_back(temp);
            return;
        }

        int n = candidates.size();
        for (int i = idx; i < n; ++i) {
            int val = candidates[i];
            if (val <= _target) {
                temp.push_back(val);
                dfs(i, _target-val, candidates, temp, res);
                temp.pop_back();
            }
        }
        
    }
};